## 4: Creating Functions

# Table of Contents

1. **Basics of Creating Functions**  
2. **Default Values of Function Arguments**  
3. **Returning Values from a Function**  
4. **Functions with an Undefined Number of Arguments**  
5. **Recursive Functions**


1. Basics of Creating Functions
Functions in every programming language are used to logically divide longer code into simpler parts, which become easier to maintain and understand.
2. In other words, a function is a mini-program whose execution may (but does not have to) depend on input data (function arguments). A major advantage of functions is that they can be called multiple times without rewriting all the code contained within them.
3. Functions in Python are objects, so we define them using the keyword <code>def</code> function_name(arguments):
4. The keyword <code>def</code> defines a function in Python, and in the parentheses following the function name we can provide a list of input data (arguments) needed to call it, which may affect the function’s result. A function may have no arguments (as in the previous example), but then its result will always be the same. Parentheses after the function name are required to define it correctly, even if the function takes no arguments.
5. The function’s code is located in a block that is indented below the function header (after def). Curly braces {} are not used in Python to define the scope of a function (they are used in other languages like C or JavaScript). In Python, indentation determines the function’s scope. If the function contains only a single statement, indentation is not strictly necessary.
5. We call a function by providing its name and any required arguments (if needed).

In [1]:
def function_1():
    return 10

In [2]:
function_1()

10

Let’s create a function that calculates the quotient of two numbers.

In [3]:
# Definition of the function quotient
def quotient(a, b):
    return a / b

# Function call
print(quotient(1, 4))        # Returns 0.25
print(quotient(-1.42, 5e-1)) # Returns -2.84


0.25
-2.84


<p data-start="0" data-end="12">Explanation:</p>
<p data-start="14" data-end="314" data-is-last-node="" data-is-only-node=""><code data-start="14" data-end="33">def quotient(a, b):</code> &mdash; We define a function named <strong data-start="63" data-end="73" data-is-only-node="">iloraz</strong> (quotient) that takes two arguments: <code data-start="111" data-end="114">a</code> and <code data-start="119" data-end="122">b</code>.<br data-start="123" data-end="126" /><code data-start="126" data-end="140">return a / b</code> &mdash; The function returns the result of dividing <code data-start="187" data-end="190">a</code> by <code data-start="194" data-end="197">b</code>.<br data-start="198" data-end="201" /><strong data-start="201" data-end="220">Function calls:</strong> The function is called with different arguments, and the result is displayed using <code data-start="304" data-end="313">print()</code>.</p>

<p data-start="0" data-end="8"><strong data-start="0" data-end="8">NOTE</strong></p>
<p data-start="10" data-end="113" data-is-last-node="" data-is-only-node="">In Python, we use <code data-start="28" data-end="33">def</code> to define a function.<br data-start="55" data-end="58" />A function returns a result using the keyword <code data-start="104" data-end="112">return</code>.</p>

When calling a function without specifying the argument names, we must provide their values in the correct order — exactly as they are listed in the function definition.
In this case, the value for a should be given first, followed by b.
If we want to provide them in a different order, we must use their names when calling the function.

In [4]:
print(iloraz(5, -2))
print(iloraz(a = 5, b = -2))
print(iloraz(b = 5, a = -2))

-2.5
-2.5
-0.4


<p data-start="100" data-end="167">These lines call the <strong data-start="121" data-end="133">quotient</strong> function in three different ways:</p>
<ol data-start="169" data-end="495" data-is-last-node="" data-is-only-node="">
<li data-start="169" data-end="251">
<p data-start="172" data-end="251"><code data-start="172" data-end="196">print(quotient(5, -2))</code> &mdash; Calls the function using <strong data-start="224" data-end="248">positional arguments</strong>.</p>
</li>
<li data-start="252" data-end="365">
<p data-start="255" data-end="365"><code data-start="255" data-end="287">print(quotient(a = 5, b = -2))</code> &mdash; Calls the function using <strong data-start="315" data-end="344">named (keyword) arguments</strong> in the same order.</p>
</li>
<li data-start="366" data-end="495" data-is-last-node="">
<p data-start="369" data-end="495" data-is-last-node=""><code data-start="369" data-end="401">print(quotient(b = 5, a = -2))</code> &mdash; Calls the function using <strong data-start="429" data-end="448">named arguments</strong> in a <strong data-start="454" data-end="473" data-is-only-node="">different order</strong>, which is also valid.</p>
</li>
</ol>

<h2 data-start="0" data-end="42">2. Default Argument Values in Functions</h2>
<p data-start="44" data-end="309">Function arguments can have <strong data-start="72" data-end="90">default values</strong> assigned to them. In that case, using that argument when calling the function becomes <strong data-start="177" data-end="189">optional</strong>.<br data-start="190" data-end="193" />If we call the function without assigning a value to such an argument, the <strong data-start="268" data-end="285">default value</strong> will be used instead.</p>
<p data-start="311" data-end="500" data-is-last-node="" data-is-only-node="">Remember that when calling a function, you must assign values to all arguments <strong data-start="390" data-end="433">that do not have default values defined</strong>.<br data-start="434" data-end="437" />If you don&rsquo;t, <strong data-start="451" data-end="461">Python</strong> (not R) will display an error message.</p>

In [4]:
def quotient2(a=None, b=100):
    if a is None:
        raise ValueError("The argument 'a' must be provided")
    return a / b


print(quotient2(20))       # Uses the default value of b (100)
print(quotient2(20, 50))   # b is explicitly set to 50


0.2
0.4


<h2 data-start="0" data-end="38">3. Returning Values from a Function</h2>
<p data-start="40" data-end="273" data-is-last-node="" data-is-only-node="">To make it clear what a function returns, it is generally recommended to use the <code data-start="121" data-end="141">return(expression)</code> statement.<br data-start="152" data-end="155" />The <code data-start="159" data-end="169">return()</code> command also allows you to control which expression (or object) will be the <strong data-start="246" data-end="256">output</strong> of the function.</p>

In [5]:
def quotient4(a, b=100):
    result = a / b
    i = 0  # This line does nothing because `i` is not used
    return result

print(quotient4(25))


0.25


<p>Using the <code data-start="10" data-end="20">return()</code> statement also allows you to vary the function&rsquo;s output depending on whether certain conditions are met.</p>

In [6]:
def number(a):
    if a > 0:
        return "positive"   # positive number
    elif a < 0:
        return "negative"   # negative number
    else:
        return "zero"       # number equal to zero

# Examples of use
print(number(10e2))   # 1000 in scientific notation
print(number(-2.3))   # Negative number
print(number(0.000))  # Number equal to zero


positive
negative
zero


<p data-start="0" data-end="16"><strong data-start="0" data-end="16">Explanation:</strong></p>
<p data-start="18" data-end="333" data-is-last-node="" data-is-only-node="">The function <code data-start="31" data-end="39">number</code> takes one argument, <code data-start="60" data-end="63">a</code>.<br data-start="64" data-end="67" />It checks whether <code data-start="85" data-end="88">a</code> is greater than 0, less than 0, or equal to 0, and returns the corresponding text: <code data-start="172" data-end="184">"positive"</code>, <code data-start="186" data-end="198">"negative"</code>, or <code data-start="203" data-end="211">"zero"</code>.<br data-start="212" data-end="215" />In Python, <code data-start="226" data-end="232">elif</code> is used for the second condition, while <code data-start="273" data-end="279">else</code> is used when none of the previous conditions are met.</p>

In [7]:
def number(a):
    if a > 0:
        result = "positive"   # positive number
    elif a < 0:
        result = "negative"   # negative number
    else:
        result = "zero"       # number equal to zero
    print(result)  # Using print instead of return


<p>Of course, we can also write this function without using the <code data-start="61" data-end="71">return()</code> statement.</p>

In [8]:
number(1)

positive


<p data-start="0" data-end="460">However, using the <code data-start="19" data-end="29">return()</code> statement directly is recommended for both <strong data-start="73" data-end="93">code readability</strong> and <strong data-start="98" data-end="112">efficiency</strong>.<br data-start="113" data-end="116" />Calling the <code data-start="128" data-end="139">number2()</code> function will always execute its entire code, whereas in the case of the <code data-start="213" data-end="223">number()</code> function, encountering the first <code data-start="257" data-end="267">return()</code> statement will terminate the function and skip the remaining code.<br data-start="334" data-end="337" />Therefore, for example, when <code data-start="366" data-end="373">a &gt; 0</code>, the <code data-start="379" data-end="389">number()</code> function will execute slightly faster than the <code data-start="437" data-end="448">number2()</code> function.</p>
<p data-start="462" data-end="623" data-is-last-node="" data-is-only-node="">A function can return only a <strong data-start="491" data-end="508">single object</strong>.<br data-start="509" data-end="512" />If we want it to return multiple objects, we can group them into a <strong data-start="579" data-end="587">list</strong> and return that list as the result.</p>

In [9]:
def quotient5(a, b=100):
    result = {
        "dividend": a,   # Corresponds to "dzielna" in R
        "divisor": b,    # Corresponds to "dzielnik" in R
        "result": a / b  # Corresponds to "wynik" in R
    }
    return result

# Examples of use
print(quotient5(25))       # Default divisor: 100
print(quotient5(20, 50))   # Provided divisor: 50


{'dividend': 25, 'divisor': 100, 'result': 0.25}
{'dividend': 20, 'divisor': 50, 'result': 0.4}


<p data-start="0" data-end="16"><strong data-start="0" data-end="16">Explanation:</strong></p>
<p data-start="18" data-end="174">The function <code data-start="31" data-end="42">quotient5</code> takes two arguments: <code data-start="64" data-end="67">a</code> (required) and <code data-start="83" data-end="86">b</code> (default value 100).<br data-start="107" data-end="110" />It creates a <strong data-start="123" data-end="146" data-is-only-node="">dictionary (<code data-start="137" data-end="143">dict</code>)</strong> that contains three keys:</p>
<ul data-start="175" data-end="293">
<li data-start="175" data-end="210">
<p data-start="177" data-end="210"><code data-start="177" data-end="189">"dividend"</code>: the value of <code data-start="204" data-end="207">a</code>,</p>
</li>
<li data-start="211" data-end="245">
<p data-start="213" data-end="245"><code data-start="213" data-end="224">"divisor"</code>: the value of <code data-start="239" data-end="242">b</code>,</p>
</li>
<li data-start="246" data-end="293">
<p data-start="248" data-end="293"><code data-start="248" data-end="258">"result"</code>: the result of dividing <code data-start="283" data-end="290">a / b</code>.</p>
</li>
</ul>
<p data-start="295" data-end="415">The function returns this dictionary using <code data-start="338" data-end="346">return</code>.<br data-start="347" data-end="350" />The results are printed in Python using the <code data-start="394" data-end="403">print()</code> function.</p>
<p data-start="417" data-end="435">Example outputs:</p>
<ul data-start="436" data-end="587" data-is-last-node="" data-is-only-node="">
<li data-start="436" data-end="511">
<p data-start="438" data-end="511">For <code data-start="442" data-end="457">quotient5(25)</code>: <code data-start="459" data-end="509">{'dividend': 25, 'divisor': 100, 'result': 0.25}</code></p>
</li>
<li data-start="512" data-end="587" data-is-last-node="">
<p data-start="514" data-end="587" data-is-last-node="">For <code data-start="518" data-end="537">quotient5(20, 50)</code>: <code data-start="539" data-end="587" data-is-last-node="">{'dividend': 20, 'divisor': 50, 'result': 0.4}</code></p>
</li>
</ul>

Argumenty funkcji można wypisać przy użyciu biblioteki `inspect`

In [11]:
import inspect

# Definicja funkcji
def iloraz5(a, b=100):
    wynik = {
        "dzielna": a,
        "dzielnik": b,
        "wynik": a / b
    }
    return wynik

# Wypisanie argumentów funkcji
print(inspect.signature(iloraz5))  

(a, b=100)


<h2 data-start="0" data-end="53">4. Functions with an Undefined Number of Arguments</h2>
<p data-start="55" data-end="369" data-is-last-node="" data-is-only-node="">In Python, we can also create functions <strong data-start="95" data-end="139">without a predefined number of arguments</strong>.<br data-start="140" data-end="143" />To do this, we include an asterisk <code data-start="178" data-end="185">*args</code> in the function definition in place of the arguments whose names and number are not specified.<br data-start="280" data-end="283" />Inside the function, we refer to this group of arguments in the same way &mdash; as <code data-start="361" data-end="368">*args</code>.</p>

In [10]:
def join_texts(*args):
    return "".join(args)  # Joins all arguments into a single string

# Examples of use
print(join_texts("k", "o", "mp", "ute", "r"))   # Returns "komputer"

print(join_texts("P", "Y", "t", "h", "o", "n")) # Returns "PYthon"


komputer
PYthon


<p data-start="0" data-end="16"><strong data-start="0" data-end="16">Explanation:</strong></p>
<p data-start="18" data-end="258" data-is-last-node="" data-is-only-node="">The function accepts any number of arguments using <code data-start="69" data-end="76">*args</code>.<br data-start="77" data-end="80" /><code data-start="80" data-end="95">"".join(args)</code> combines all the given arguments into a single string without any separator (equivalent to <code data-start="187" data-end="200">paste0(...)</code>in R).<br data-start="207" data-end="210" />The result is returned as one continuous string.</p>

We can combine defined and undefined arguments in a single function.
However, if we want to change the values of the defined arguments that appear after the undefined ones, we must specify their names when calling the function.

In [11]:
def join_texts2(*args, separator="#"):
    return separator.join(args)  # Joins the arguments using the specified separator

# Examples of use
print(join_texts2("k", "o", "mp", "ute", "r"))                 # Default separator "#"
print(join_texts2("k", "o", "mp", "ute", "r", separator=","))  # Separator ","


k#o#mp#ute#r
k,o,mp,ute,r


<p data-start="0" data-end="16"><strong data-start="0" data-end="16">Explanation:</strong></p>
<p data-start="18" data-end="319" data-is-last-node="" data-is-only-node="">The function <code data-start="31" data-end="44">join_texts2</code> accepts any number of positional arguments using <code data-start="94" data-end="101">*args</code>.<br data-start="102" data-end="105" />The optional argument <code data-start="127" data-end="138">separator</code> specifies which character should separate the elements (default is <code data-start="206" data-end="211">"#"</code>).<br data-start="213" data-end="216" />The function uses <code data-start="234" data-end="258">"separator".join(args)</code> to combine the string elements with the specified separator.</p>

<p data-start="0" data-end="214">Another example &mdash; calculating the length of a segment connecting the point <span class="katex"><span class="katex-mathml">[0,0,...,0]</span><span class="katex-html" aria-hidden="true"><span class="base"><span class="mopen">[</span><span class="mord">0</span><span class="mpunct">,</span><span class="mord">0</span><span class="mpunct">,</span><span class="mord">...</span><span class="mpunct">,</span><span class="mord">0</span><span class="mclose">]</span></span></span></span> with the point <span class="katex"><span class="katex-mathml">[x1,x2,...,xk]</span><span class="katex-html" aria-hidden="true"><span class="base"><span class="mopen">[</span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist"><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span><span class="vlist-s">​</span></span></span></span></span><span class="mpunct">,</span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist"><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span><span class="vlist-s">​</span></span></span></span></span><span class="mpunct">,</span><span class="mord">...</span><span class="mpunct">,</span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist"><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">k</span></span></span></span><span class="vlist-s">​</span></span></span></span></span><span class="mclose">]</span></span></span></span> in a <em data-start="136" data-end="139">k</em>-dimensional space, where the number of dimensions <code data-start="190" data-end="193">k</code> is not predefined.</p>
<p data-start="216" data-end="362" data-is-last-node="" data-is-only-node="">We calculate the length of the segment as the <strong data-start="262" data-end="284">Euclidean distance</strong>, that is, the square root of the sum of the squares of the coordinate values.</p>

In [12]:
import math

def segment_length(*args):
    length = 0
    for coordinate in args:
        length += coordinate ** 2  # Add the square of the coordinate
    return math.sqrt(length)       # Return the square root of the sum of squares

# Examples of use
print(segment_length(3, 4))           # For (3, 4)
print(segment_length(3, -4, 5, -6))   # For (3, -4, 5, -6)


5.0
9.273618495495704


<p data-start="0" data-end="16"><strong data-start="0" data-end="16">Explanation:</strong></p>
<ul data-start="18" data-end="295" data-is-last-node="" data-is-only-node="">
<li data-start="18" data-end="98">
<p data-start="20" data-end="98"><strong data-start="20" data-end="34">Arguments:</strong> The function accepts any number of coordinates using <code data-start="88" data-end="95" data-is-only-node="">*args</code>.</p>
</li>
<li data-start="99" data-end="206">
<p data-start="101" data-end="206"><strong data-start="101" data-end="128">Squares of coordinates:</strong> In the loop, each coordinate is squared and added to the variable <code data-start="195" data-end="203" data-is-only-node="">length</code>.</p>
</li>
<li data-start="207" data-end="295" data-is-last-node="">
<p data-start="209" data-end="295" data-is-last-node=""><strong data-start="209" data-end="225">Square root:</strong> After the loop ends, the result is square-rooted using <code data-start="281" data-end="294" data-is-only-node="">math.sqrt()</code>.</p>
</li>
</ul>
<!DOCTYPE html>
<html lang="pl">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>results: segment_length</title>
</head>
<body>
    <h2>Results:</h2>
    <ul>
        <li>
            <code>segment_length(3, 4)</code> zwróci 
            \( \sqrt{3^2 + 4^2} = 5.0 \).
        </li>
        <li>
            <code>segment_length(3, -4, 5, -6)</code> zwróci 
            \( \sqrt{3^2 + (-4)^2 + 5^2 + (-6)^2} = 9.273618495495704 \).
        </li>
    </ul>
</body>
</html>


<h2 data-start="0" data-end="25">5. Recursive Functions</h2>
<p data-start="27" data-end="84">A <strong data-start="29" data-end="51">recursive function</strong> is a function that calls itself.</p>
<p data-start="86" data-end="208">This specific programming technique is often used to solve problems by dividing them into smaller and simpler subproblems.</p>
<p data-start="210" data-end="299">One of the most well-known examples is calculating the <strong data-start="265" data-end="278">factorial</strong> of a natural number.</p>
<p data-start="301" data-end="477">The factorial of any natural number is defined as the product of all consecutive natural numbers from 1 up to that number.<br data-start="423" data-end="426" />For example, the factorial of 5 (denoted as 5!) is:</p>
<p data-start="479" data-end="507">5! = 1 &times; 2 &times; 3 &times; 4 &times; 5 = 120</p>
<p data-start="509" data-end="625">The task of finding the factorial of 5 can be broken down into a smaller task &mdash; multiplying the factorial of 4 by 5:</p>
<p data-start="627" data-end="638">5! = 5 &times; 4!</p>
<p data-start="640" data-end="675">In general, we can express this as:</p>
<p data-start="677" data-end="694">n! = n &times; (n &minus; 1)!</p>
<p data-start="696" data-end="789" data-is-last-node="" data-is-only-node="">This decomposition continues until we reach the <strong data-start="744" data-end="757">base case</strong>: 0!, which is defined as <strong data-start="783" data-end="788">1</strong>.</p>

In [13]:
def factorial(n):
    if n == 0:  # Base case
        return 1
    else:        # Recursive call
        return n * factorial(n - 1)

# Example of use
print(factorial(5))  # Result: 120


120


<p data-start="0" data-end="16"><strong data-start="0" data-end="16">Explanation:</strong></p>
<ul data-start="18" data-end="220" data-is-last-node="" data-is-only-node="">
<li data-start="18" data-end="106">
<p data-start="20" data-end="106"><strong data-start="20" data-end="39">Base condition:</strong> If <code data-start="43" data-end="51" data-is-only-node="">n == 0</code>, the function returns <code data-start="74" data-end="77">1</code>, which ends the recursion.</p>
</li>
<li data-start="107" data-end="220" data-is-last-node="">
<p data-start="109" data-end="220" data-is-last-node=""><strong data-start="109" data-end="123">Recursion:</strong> If <code data-start="127" data-end="134" data-is-only-node="">n &gt; 0</code>, the function calls itself with the argument <code data-start="180" data-end="187">n - 1</code> until the base condition is met.</p>
</li>
</ul>
<li>Results:
<ul>
<li>For <code>factorial(5)</code>:
<ul>
<li><code>5 * factorial(4)</code></li>
<li><code>5 * (4 * factorial(3))</code></li>
<li><code>5 * (4 * (3 * factorial(2)))</code></li>
<li><code>5 * (4 * (3 * (2 * factorial(1))))</code></li>
<li><code>5 * (4 * (3 * (2 * (1 * factorial(0)))))</code></li>
<li><code>5 * 4 * 3 * 2 * 1 = 120</code></li>
</ul>
</li>
</ul>
</li>
</ol>

Using recursion makes the code in this case short and elegant.
However, because of these nested calls, it can sometimes be harder to understand the logic of the function.
Recursive functions are also more resource-intensive (they use more computer memory) because they can create many nested function calls.
This is something to keep in mind when using recursion to solve large problems.
We will return to the topic of recursion and its optimization later in this course.

### 6. Writing mean in a fucntion suing loop

In [5]:
import numpy as np
def mean_custom(vector = None):
    if vector is not None:
        x = vector[0]
        for i in range(1,len(vector)):
            x = x + vector[i]
    
        return x/len(vector)
    else:
        return print("None vector")

    
print(mean_custom(vector = [1,2,3]))   
print(np.mean( [1,2,3]))

print(mean_custom(vector = None))   
# print(np.mean(None))



2.0
2.0
None vector
None


In [6]:
def minmax_custom(vector=None):
    if vector is None or len(vector) == 0:
        print("Empty or None vector.")
        return None
    min_val, max_val = vector[0], vector[0]
    for v in vector:
        if v < min_val:
            min_val = v
        if v > max_val:
            max_val = v
    return (min_val, max_val)

print(minmax_custom([5, 2, 9, -1, 0]))
print(min([5, 2, 9, -1, 0]), max([5, 2, 9, -1, 0]))


(-1, 9)
-1 9


In [7]:
def median_custom(vector=None):
    if vector is None or len(vector) == 0:
        print("Empty or None vector.")
        return None
    vector_sorted = sorted(vector)
    n = len(vector_sorted)
    mid = n // 2
    if n % 2 == 0:
        return (vector_sorted[mid - 1] + vector_sorted[mid]) / 2
    else:
        return vector_sorted[mid]

print(median_custom([5, 3, 1, 4, 2]))
print(np.median([5, 3, 1, 4, 2]))


3
3.0


<h2 style="text-align: center;">Time for exercie :)</h2>

### Exercise 1: Coefficient of Variation
Task:
Write a function that, based on a vector of numerical values, calculates the coefficient of variation.

### Exercise 2: Cumulative Sum
Task:
Write a function that calculates the cumulative sum of any number of arguments.
Compare the results with the cumsum() function.

### Exercise 3: Minkowski Distance
Task:
Modify the segment_length() function so that it calculates the Minkowski distance, that is, the m-th root of the sum of the absolute values of the coordinates raised to the power of m.
The previously calculated Euclidean distance is a special case of the Minkowski distance for m = 2.

### Exercise 4: Fibonacci Sequence
Task:
Write a function that, for any given value of n, displays the first n Fibonacci numbers.

### Exercise 5: Prime Number
Task:
A prime number is a number that is divisible (without a remainder) only by 1 and itself.
Write a function that checks whether a given natural number is a prime number.

### Exercise 6: Binary System
Task:
Write a function that, for a given natural number, returns a vector containing the representation of that number in the binary system.

### Exercise 1: Plots
Write a function that works with any dataset (DataFrame) and, for each variable, performs the following:
Integer variable → create a boxplot
Float (numeric) variable → create a histogram
Categorical variable → create a bar plot
You can use the mpg dataset for testing.

### Exercise 1: Extended Coefficient of Variation
Write a function that calculates the coefficient of variation.
Modify the function to include the following features:
a) A parameter that lets the user decide whether the result should be returned as a fraction or in percentage points.
b) The ability to control whether NA/missing values should be included in the calculations.
Extend the function to allow computing the coefficient of variation for subsamples grouped by a nominal (categorical) variable.